File: player-core/src/main/java/com/libreshockwave/player/frame/FrameContext.java
package com.libreshockwave.player.frame;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.chunks.ScoreChunk;
import com.libreshockwave.player.PlayerEvent;
import com.libreshockwave.player.behavior.BehaviorInstance;
import com.libreshockwave.player.behavior.BehaviorManager;
import com.libreshockwave.player.event.EventDispatcher;
import com.libreshockwave.player.render.SpriteRegistry;
import com.libreshockwave.player.score.ScoreBehaviorRef;
import com.libreshockwave.player.score.ScoreNavigator;
import com.libreshockwave.player.score.SpriteSpan;
import com.libreshockwave.player.sprite.SpriteState;
import com.libreshockwave.player.timeout.TimeoutManager;
import com.libreshockwave.util.ValueProvider;
import com.libreshockwave.vm.datum.Datum;
import com.libreshockwave.vm.LingoVM;
import com.libreshockwave.vm.util.AncestorChainWalker;

import java.util.*;
import java.util.function.Consumer;

/**
 * Manages the execution context for a single frame.
 * Handles sprite initialization, event dispatch, and frame transitions.
 */
public class FrameContext {

    private final DirectorFile file;
    private final LingoVM vm;
    private final ScoreNavigator navigator;
    private final BehaviorManager behaviorManager;
    private final EventDispatcher eventDispatcher;

    private TimeoutManager timeoutManager;  // Set by Player for system event forwarding
    private ValueProvider<Datum> actorListSupplier;  // Provides _movie.actorList
    private SpriteRegistry spriteRegistry;  // For checking puppet state during frame transitions

    private int currentFrame = 1;
    private Integer pendingFrame = null;  // Set by go/jump commands
    private boolean inFrameScript = false;

    // Active sprite channels
    private final Set<Integer> activeChannels = new HashSet<>();
    private final Set<Integer> enteredChannels = new LinkedHashSet<>();

    // Debug logging
    private boolean debugEnabled = false;
    private Consumer<FrameEvent> eventListener;

    public FrameContext(DirectorFile file, LingoVM vm) {
        this.file = file;
        this.vm = vm;
        this.navigator = new ScoreNavigator(file);
        this.behaviorManager = new BehaviorManager(file);
        this.eventDispatcher = new EventDispatcher(file, vm, behaviorManager);
    }

    public void setDebugEnabled(boolean enabled) {
        this.debugEnabled = enabled;
        behaviorManager.setDebugEnabled(enabled);
        eventDispatcher.setDebugEnabled(enabled);
    }

    public void setEventListener(Consumer<FrameEvent> listener) {
        this.eventListener = listener;
    }

    public void setTimeoutManager(TimeoutManager timeoutManager) {
        this.timeoutManager = timeoutManager;
    }

    public void setActorListSupplier(ValueProvider<Datum> supplier) {
        this.actorListSupplier = supplier;
    }

    public void setSpriteRegistry(SpriteRegistry registry) {
        this.spriteRegistry = registry;
    }

    public int getCurrentFrame() {
        return currentFrame;
    }

    public int getEffectiveFrame() {
        return pendingFrame != null ? pendingFrame : currentFrame;
    }

    public int getFrameCount() {
        return navigator.getFrameCount();
    }

    public ScoreNavigator getNavigator() {
        return navigator;
    }

    public BehaviorManager getBehaviorManager() {
        return behaviorManager;
    }

    public EventDispatcher getEventDispatcher() {
        return eventDispatcher;
    }

    // Frame navigation

    /**
     * Go to a specific frame (queued for next advance).
     */
    public void goToFrame(int frame) {
        int max = getFrameCount();
        if (frame >= 1 && frame <= max) {
            pendingFrame = frame;
        }
    }

    /**
     * Force an immediate frame transition, bypassing the exitFrame handler.
     * Used when external navigation (e.g. mouse click → go to "label") must
     * override an exitFrame script that loops with "go to the frame".
     * In Director, user-initiated go() during mouse events takes priority.
     */
    public void forceGoToFrame(int frame) {
        int max = getFrameCount();
        if (frame < 1 || frame > max) return;

        int oldFrame = currentFrame;
        logEvent("forceGoToFrame: " + oldFrame + " -> " + frame);

        endSpritesLeavingFrame(oldFrame, frame);
        behaviorManager.clearFrameScript();

        currentFrame = frame;
        pendingFrame = null;  // Clear any pending to prevent exitFrame override
        enterFrame(frame);
    }

    /**
     * Go to a labeled frame.
     */
    public void goToLabel(String label) {
        int frame = navigator.getFrameForLabel(label);
        if (frame > 0) {
            goToFrame(frame);
        }
    }

    // Frame execution

    /**
     * Initialize the first frame (called on movie start).
     */
    public void initializeFirstFrame() {
        currentFrame = 1;
        pendingFrame = null;
        activeChannels.clear();
        enteredChannels.clear();
        behaviorManager.clear();

        logEvent("initializeFirstFrame");

        // Begin sprites for frame 1
        beginSpritesForFrame(currentFrame);

        // Create frame script instance if any
        initializeFrameScript(currentFrame);
    }

    /**
     * Execute one frame update cycle (for frame loop, not first frame).
     * First frame events are handled in Player.prepareMovie().
     * Returns true if frame was executed successfully.
     */
    public boolean executeFrame() {
        return executeFrame(false);
    }

    public boolean executeFrame(boolean suppressSpriteEnterFrameOnce) {
        List<Datum.ScriptInstance> actorSnapshot = getActorSnapshot();

        // 1. stepFrame -> actorList, then behaviors + frame/movie scripts
        dispatchToActorList(actorSnapshot, "stepFrame");
        dispatchEvent(PlayerEvent.STEP_FRAME);

        // 2. prepareFrame -> actorList, timeout targets, then behaviors + frame/movie scripts
        dispatchToActorList(actorSnapshot, "prepareFrame");
        if (timeoutManager != null) {
            timeoutManager.dispatchSystemEvent(vm, "prepareFrame");
        }
        dispatchEvent(PlayerEvent.PREPARE_FRAME);

        // 3. enterFrame -> actorList, then behaviors + frame/movie scripts
        dispatchToActorList(actorSnapshot, "enterFrame");
        inFrameScript = true;
        if (suppressSpriteEnterFrameOnce) {
            eventDispatcher.dispatchFrameAndMovieEvent(PlayerEvent.ENTER_FRAME, List.of());
            notifyEvent(PlayerEvent.ENTER_FRAME);
        } else {
            dispatchEvent(PlayerEvent.ENTER_FRAME);
        }
        inFrameScript = false;

        return true;
    }

    public boolean hasFrameOneFrameScriptAndFrameTwoSpriteBehaviorStartupShape() {
        if (navigator.getFrameScript(1) == null) {
            return false;
        }
        return !hasScoreSpriteBehaviors(1) && hasScoreSpriteBehaviors(2);
    }

    /**
     * Dispatch beginSprite events to newly entered sprites.
     * Called from Player.prepareMovie() for first frame setup.
     */
    public void dispatchBeginSpriteEvents() {
        dispatchBeginSprite();
    }


    /**
     * External casts can finish loading after a score sprite has already entered.
     * Rebind any authored behaviors from that cast onto the current frame's live sprites.
     */
    public void rebindBehaviorsForLoadedCast(int castLibNumber) {
        if (castLibNumber <= 0) {
            return;
        }

        for (SpriteSpan span : navigator.getActiveSprites(currentFrame)) {
            int channel = span.getChannel();
            if (!activeChannels.contains(channel)) {
                continue;
            }
            for (ScoreBehaviorRef behaviorRef : span.getBehaviors()) {
                if (behaviorRef.castLib() != castLibNumber
                        || behaviorManager.hasInstanceForChannel(channel, behaviorRef)) {
                    continue;
                }
                if (spriteRegistry != null) {
                    spriteRegistry.markScoreBehaviorChannel(channel);
                }
                BehaviorInstance instance = behaviorManager.createInstance(behaviorRef, channel);
                if (instance == null) {
                    continue;
                }
                eventDispatcher.dispatchBehaviorEvent(instance, PlayerEvent.BEGIN_SPRITE, List.of());
                instance.setBeginSpriteCalled(true);
                logEvent("rebindBehavior: channel " + channel + " -> " + behaviorRef);
            }
        }

        ScoreBehaviorRef frameScript = navigator.getFrameScript(currentFrame);
        if (frameScript == null
                || frameScript.castLib() != castLibNumber
                || behaviorManager.getFrameScriptInstance() != null) {
            return;
        }

        BehaviorInstance frameInstance = behaviorManager.getOrCreateFrameScript(frameScript, currentFrame);
        if (frameInstance != null) {
            eventDispatcher.dispatchBehaviorEvent(frameInstance, PlayerEvent.BEGIN_SPRITE, List.of());
            frameInstance.setBeginSpriteCalled(true);
            logEvent("rebindFrameScript: frame " + currentFrame + " -> " + frameScript);
        }
    }

    /**
     * Advance to the next frame (or pending frame if set).
     * Returns the new frame number.
     *
     * Matching dirplayer-rs: exitFrame events fire BEFORE computing the destination,
     * so that go(the frame) during exitFrame correctly sets pendingFrame for this cycle.
     */
    public int advanceFrame() {
        int oldFrame = currentFrame;

        // 1. Dispatch exitFrame events FIRST (scripts may call go() here)
        List<Datum.ScriptInstance> actorSnapshot = getActorSnapshot();
        dispatchToActorList(actorSnapshot, "exitFrame");
        if (timeoutManager != null) {
            timeoutManager.dispatchSystemEvent(vm, "exitFrame");
        }
        dispatchEvent(PlayerEvent.EXIT_FRAME);

        // 2. NOW decide destination (picks up pendingFrame set by go() during exitFrame)
        int newFrame;
        if (pendingFrame != null) {
            newFrame = pendingFrame;
            pendingFrame = null;
        } else {
            newFrame = currentFrame + 1;
        }

        int max = getFrameCount();
        if (newFrame > max) {
            newFrame = 1;  // Loop
        }

        // 3. Transition sprites and enter new frame
        if (newFrame != oldFrame) {
            logEvent("advanceFrame: " + oldFrame + " -> " + newFrame);

            endSpritesLeavingFrame(oldFrame, newFrame);
            behaviorManager.clearFrameScript();

            currentFrame = newFrame;
            enterFrame(newFrame);
        }

        return currentFrame;
    }

    /**
     * Enter a new frame (initialization).
     */
    private void enterFrame(int frame) {
        // Begin new sprites
        beginSpritesForFrame(frame);

        // Initialize frame script
        initializeFrameScript(frame);

        // Dispatch beginSprite events to newly entered sprites
        dispatchBeginSprite();

        // Clear entered set for next beginSprite dispatch
        enteredChannels.clear();
    }

    // Sprite management

    /**
     * Begin sprites that are active in the given frame.
     */
    private void beginSpritesForFrame(int frame) {
        List<SpriteSpan> spans = navigator.getActiveSprites(frame);
        spans.sort(Comparator.comparingInt(SpriteSpan::getChannel));
        for (SpriteSpan span : spans) {
            int channel = span.getChannel();

            if (!activeChannels.contains(channel)) {
                // New sprite entering
                activeChannels.add(channel);
                enteredChannels.add(channel);
                ensureScoreSpriteState(frame, channel);

                // Create behavior instances for this sprite
                for (ScoreBehaviorRef behaviorRef : span.getBehaviors()) {
                    if (spriteRegistry != null) {
                        spriteRegistry.markScoreBehaviorChannel(channel);
                    }
                    behaviorManager.createInstance(behaviorRef, channel);
                }

                logEvent("beginSprite: channel " + channel);
            }
        }
    }

    /**
     * Director exposes the score-authored sprite state before beginSprite runs.
     * Seed the runtime registry from the current frame's score entry so first-touch
     * behavior reads/writes mutate the authored sprite rather than a blank puppet shell.
     */
    private void ensureScoreSpriteState(int frame, int channel) {
        if (spriteRegistry == null || file == null) {
            return;
        }
        ScoreChunk score = file.getScoreChunk();
        if (score == null) {
            return;
        }
        int frameIndex = frame - 1;
        for (ScoreChunk.FrameChannelEntry entry : score.frameData().frameChannelData()) {
            if (entry.frameIndex().value() == frameIndex
                    && entry.channelIndex().value() == channel) {
                spriteRegistry.getOrCreate(channel, entry.data());
                return;
            }
        }
    }

    /**
     * End sprites that are leaving when transitioning frames.
     * In Director, puppeted sprites persist across frame transitions —
     * they are NOT removed when their Score span ends.
     */
    private void endSpritesLeavingFrame(int oldFrame, int newFrame) {
        Set<Integer> newActiveChannels = navigator.getActiveChannels(newFrame);

        List<Integer> leaving = new ArrayList<>();
        for (int channel : activeChannels) {
            if (!newActiveChannels.contains(channel)) {
                // Check if this sprite is puppeted — puppeted sprites persist
                if (spriteRegistry != null) {
                    SpriteState state = spriteRegistry.get(channel);
                    if (state != null && state.isPuppet()) {
                        logEvent("endSprite SKIPPED (puppeted): channel " + channel);
                        continue;
                    }
                }
                leaving.add(channel);
            }
        }

        for (int channel : leaving) {
            // Dispatch endSprite event
            eventDispatcher.dispatchSpriteEvent(channel, PlayerEvent.END_SPRITE, List.of());

            // Mark instances as ended
            for (BehaviorInstance instance : behaviorManager.getInstancesForChannel(channel)) {
                instance.setEndSpriteCalled(true);
            }

            // Remove from active set
            activeChannels.remove(channel);
            behaviorManager.removeInstancesForChannel(channel);
            if (spriteRegistry != null) {
                spriteRegistry.remove(channel);
            }

            logEvent("endSprite: channel " + channel);
        }
    }


    /**
     * Initialize the frame script for the given frame.
     */
    private void initializeFrameScript(int frame) {
        ScoreBehaviorRef frameScript = navigator.getFrameScript(frame);
        if (frameScript != null) {
            behaviorManager.getOrCreateFrameScript(frameScript, frame);
            logEvent("initializeFrameScript: frame " + frame + " -> " + frameScript);
        } else {
            logEvent("initializeFrameScript: frame " + frame + " has no frame script");
        }


    }

    /**
     * Get a snapshot of the current actorList for frame event dispatch.
     */
    private List<Datum.ScriptInstance> getActorSnapshot() {
        if (actorListSupplier == null) return List.of();
        Datum list = actorListSupplier.get();
        if (!(list instanceof Datum.List actors)) return List.of();
        List<Datum.ScriptInstance> result = new ArrayList<>();
        for (Datum item : actors.items()) {
            if (item instanceof Datum.ScriptInstance si) result.add(si);
        }
        return result;
    }

    /**
     * Dispatch a frame event to all actorList members by walking each object's ancestor chain.
     * Director broadcasts all frame events (stepFrame, prepareFrame, enterFrame, exitFrame)
     * to actorList members, not just stepFrame.
     */
    private void dispatchToActorList(List<Datum.ScriptInstance> actors, String handlerName) {
        if (actors.isEmpty()) return;

        for (Datum.ScriptInstance instance : actors) {
            try {
                AncestorChainWalker.invokeHandler(vm, instance, handlerName, List.of(instance));
            } catch (Exception ignored) {
                // Silently skip errors in actorList event dispatch
            }
        }
    }

    // Event dispatch

    /**
     * Dispatch a player event.
     */
    private void dispatchEvent(PlayerEvent event) {
        eventDispatcher.dispatchGlobalEvent(event, List.of());
        notifyEvent(event);
    }

    /**
     * Dispatch beginSprite to newly entered sprites.
     */
    private void dispatchBeginSprite() {
        boolean v1StartupShape = hasFrameOneFrameScriptAndFrameTwoSpriteBehaviorStartupShape();
        for (int channel : enteredChannels) {
            List<BehaviorInstance> instances = behaviorManager.getInstancesForChannel(channel);
            for (BehaviorInstance instance : instances) {
                if (!instance.isBeginSpriteCalled()) {
                    eventDispatcher.dispatchBehaviorEvent(instance, PlayerEvent.BEGIN_SPRITE, List.of());
                    maybeDelayFlagsStartupCycle(instance, v1StartupShape);
                    instance.setBeginSpriteCalled(true);
                }
            }
        }

        // Also dispatch beginSprite for frame behavior if new
        BehaviorInstance frameInstance = behaviorManager.getFrameScriptInstance();
        if (frameInstance != null && !frameInstance.isBeginSpriteCalled()) {
            eventDispatcher.dispatchFrameAndMovieEvent(PlayerEvent.BEGIN_SPRITE, List.of());
            frameInstance.setBeginSpriteCalled(true);
        }
    }

    private void maybeDelayFlagsStartupCycle(BehaviorInstance instance, boolean v1StartupShape) {
        if (!v1StartupShape || instance.getScript() == null) {
            return;
        }
        if (!"Flags behavior".equals(instance.getScript().getScriptName())) {
            return;
        }
        Datum wait = instance.getProperty("pWait");
        if (wait instanceof Datum.Int i && i.value() == 3) {
            instance.setProperty("pWait", Datum.of(7));
        }
    }

    private boolean hasScoreSpriteBehaviors(int frame) {
        for (SpriteSpan span : navigator.getActiveSprites(frame)) {
            if (!span.getBehaviors().isEmpty()) {
                return true;
            }
        }
        return false;
    }

    // Debug and notification

    private void logEvent(String message) {
        if (debugEnabled) {
            System.out.println("[FrameContext] " + message);
        }
    }

    private void notifyEvent(PlayerEvent event) {
        if (eventListener != null) {
            eventListener.accept(new FrameEvent(event, currentFrame));
        }
    }

    /**
     * Reset the context (called on stop).
     */
    public void reset() {
        currentFrame = 1;
        pendingFrame = null;
        activeChannels.clear();
        enteredChannels.clear();
        behaviorManager.clear();
        inFrameScript = false;
    }

}

File: cpp/src/player/frame/FrameContext.cpp
#include "libreshockwave/player/frame/FrameContext.hpp"

#include <algorithm>
#include <iostream>
#include <stdexcept>
#include <utility>

#include "libreshockwave/DirectorFile.hpp"
#include "libreshockwave/chunks/ScoreChunk.hpp"
#include "libreshockwave/player/behavior/BehaviorInstance.hpp"
#include "libreshockwave/player/behavior/BehaviorManager.hpp"
#include "libreshockwave/player/event/EventDispatcher.hpp"
#include "libreshockwave/player/render/SpriteRegistry.hpp"
#include "libreshockwave/player/score/ScoreNavigator.hpp"
#include "libreshockwave/player/score/SpriteSpan.hpp"
#include "libreshockwave/player/sprite/SpriteState.hpp"

namespace libreshockwave::player::frame {
namespace {

template <typename T>
T& requireDependency(T* value, const char* name) {
    if (value == nullptr) {
        throw std::runtime_error(std::string("FrameContext missing dependency: ") + name);
    }
    return *value;
}

} // namespace

FrameContext::FrameContext(DirectorFile* file)
    : file_(file),
      ownedNavigator_(std::make_unique<score::ScoreNavigator>(file)),
      navigator_(ownedNavigator_.get()),
      ownedBehaviorManager_(std::make_unique<behavior::BehaviorManager>(file)),
      behaviorManager_(ownedBehaviorManager_.get()),
      ownedEventDispatcher_(std::make_unique<event::EventDispatcher>(behaviorManager_)),
      eventDispatcher_(ownedEventDispatcher_.get()) {
    bindDispatcher();
}

FrameContext::FrameContext(DirectorFile* file,
                           score::ScoreNavigator* navigator,
                           behavior::BehaviorManager* behaviorManager,
                           event::EventDispatcher* eventDispatcher)
    : file_(file),
      navigator_(navigator),
      behaviorManager_(behaviorManager),
      eventDispatcher_(eventDispatcher) {
    requireDependency(navigator_, "navigator");
    requireDependency(behaviorManager_, "behaviorManager");
    requireDependency(eventDispatcher_, "eventDispatcher");
    bindDispatcher();
}

FrameContext::~FrameContext() = default;
FrameContext::FrameContext(FrameContext&&) noexcept = default;
FrameContext& FrameContext::operator=(FrameContext&&) noexcept = default;

void FrameContext::setDebugEnabled(bool enabled) {
    debugEnabled_ = enabled;
    behaviorManager().setDebugEnabled(enabled);
    eventDispatcher().setDebugEnabled(enabled);
}

bool FrameContext::debugEnabled() const {
    return debugEnabled_;
}

void FrameContext::setEventListener(EventListener listener) {
    eventListener_ = std::move(listener);
}

void FrameContext::setActorListDispatcher(ActorListDispatcher dispatcher) {
    actorListDispatcher_ = std::move(dispatcher);
}

void FrameContext::setTimeoutEventDispatcher(TimeoutEventDispatcher dispatcher) {
    timeoutEventDispatcher_ = std::move(dispatcher);
}

void FrameContext::setSpriteRegistry(render::SpriteRegistry* registry) {
    spriteRegistry_ = registry;
    eventDispatcher().setSpriteRegistry(registry);
}

int FrameContext::currentFrame() const { return currentFrame_; }
int FrameContext::effectiveFrame() const { return pendingFrame_ > 0 ? pendingFrame_ : currentFrame_; }
int FrameContext::frameCount() const { return navigator().getFrameCount(); }
score::ScoreNavigator& FrameContext::navigator() { return requireDependency(navigator_, "navigator"); }
const score::ScoreNavigator& FrameContext::navigator() const { return requireDependency(navigator_, "navigator"); }
behavior::BehaviorManager& FrameContext::behaviorManager() { return requireDependency(behaviorManager_, "behaviorManager"); }
const behavior::BehaviorManager& FrameContext::behaviorManager() const { return requireDependency(behaviorManager_, "behaviorManager"); }
event::EventDispatcher& FrameContext::eventDispatcher() { return requireDependency(eventDispatcher_, "eventDispatcher"); }
const event::EventDispatcher& FrameContext::eventDispatcher() const { return requireDependency(eventDispatcher_, "eventDispatcher"); }
const std::set<int>& FrameContext::activeChannels() const { return activeChannels_; }
bool FrameContext::inFrameScript() const { return inFrameScript_; }

void FrameContext::goToFrame(int frame) {
    const int max = frameCount();
    if (frame >= 1 && frame <= max) {
        pendingFrame_ = frame;
    }
}

void FrameContext::forceGoToFrame(int frame) {
    const int max = frameCount();
    if (frame < 1 || frame > max) {
        return;
    }

    const int oldFrame = currentFrame_;
    logEvent("forceGoToFrame");
    endSpritesLeavingFrame(oldFrame, frame);
    behaviorManager().clearFrameScript();

    currentFrame_ = frame;
    pendingFrame_ = 0;
    enterFrame(frame);
}

void FrameContext::goToLabel(std::string_view label) {
    const int frame = navigator().getFrameForLabel(label);
    if (frame > 0) {
        goToFrame(frame);
    }
}

void FrameContext::initializeFirstFrame() {
    currentFrame_ = 1;
    pendingFrame_ = 0;
    activeChannels_.clear();
    enteredChannels_.clear();
    behaviorManager().clear();

    logEvent("initializeFirstFrame");
    beginSpritesForFrame(currentFrame_);
    initializeFrameScript(currentFrame_);
}

bool FrameContext::executeFrame(bool suppressSpriteEnterFrameOnce) {
    dispatchToActorList("stepFrame");
    dispatchEvent(PlayerEvent::StepFrame);

    dispatchToActorList("prepareFrame");
    if (timeoutEventDispatcher_) {
        timeoutEventDispatcher_("prepareFrame");
    }
    dispatchEvent(PlayerEvent::PrepareFrame);

    dispatchToActorList("enterFrame");
    inFrameScript_ = true;
    if (suppressSpriteEnterFrameOnce) {
        eventDispatcher().dispatchFrameAndMovieEvent(PlayerEvent::EnterFrame);
        notifyEvent(PlayerEvent::EnterFrame);
    } else {
        dispatchEvent(PlayerEvent::EnterFrame);
    }
    inFrameScript_ = false;
    return true;
}

void FrameContext::dispatchBeginSpriteEvents() {
    dispatchBeginSprite();
}

void FrameContext::rebindBehaviorsForLoadedCast(int castLibNumber) {
    if (castLibNumber <= 0) {
        return;
    }

    for (const auto& span : navigator().getActiveSprites(currentFrame_)) {
        const int channel = span.channel();
        if (!activeChannels_.contains(channel)) {
            continue;
        }
        for (const auto& behaviorRef : span.behaviors()) {
            if (behaviorRef.castLib() != castLibNumber ||
                behaviorManager().hasInstanceForChannel(channel, behaviorRef)) {
                continue;
            }
            if (spriteRegistry_ != nullptr) {
                spriteRegistry_->markScoreBehaviorChannel(channel);
            }
            auto instance = behaviorManager().createInstance(behaviorRef, channel);
            if (!instance) {
                continue;
            }
            eventDispatcher().dispatchBehaviorEvent(instance, PlayerEvent::BeginSprite);
            instance->setBeginSpriteCalled(true);
            logEvent("rebindBehavior");
        }
    }

    const auto* frameScript = navigator().getFrameScript(currentFrame_);
    if (frameScript == nullptr ||
        frameScript->castLib() != castLibNumber ||
        behaviorManager().frameScriptInstance() != nullptr) {
        return;
    }

    auto frameInstance = behaviorManager().getOrCreateFrameScript(*frameScript, currentFrame_);
    if (frameInstance) {
        eventDispatcher().dispatchBehaviorEvent(frameInstance, PlayerEvent::BeginSprite);
        frameInstance->setBeginSpriteCalled(true);
        logEvent("rebindFrameScript");
    }
}

int FrameContext::advanceFrame() {
    const int oldFrame = currentFrame_;

    dispatchToActorList("exitFrame");
    if (timeoutEventDispatcher_) {
        timeoutEventDispatcher_("exitFrame");
    }
    dispatchEvent(PlayerEvent::ExitFrame);

    int newFrame = 0;
    if (pendingFrame_ > 0) {
        newFrame = pendingFrame_;
        pendingFrame_ = 0;
    } else {
        newFrame = currentFrame_ + 1;
    }

    const int max = frameCount();
    if (max > 0 && newFrame > max) {
        newFrame = 1;
    }

    if (newFrame != oldFrame) {
        logEvent("advanceFrame");
        endSpritesLeavingFrame(oldFrame, newFrame);
        behaviorManager().clearFrameScript();
        currentFrame_ = newFrame;
        enterFrame(newFrame);
    }

    return currentFrame_;
}

void FrameContext::reset() {
    currentFrame_ = 1;
    pendingFrame_ = 0;
    activeChannels_.clear();
    enteredChannels_.clear();
    behaviorManager().clear();
    inFrameScript_ = false;
}

void FrameContext::bindDispatcher() {
    if (eventDispatcher_ != nullptr) {
        eventDispatcher_->setBehaviorManager(behaviorManager_);
        eventDispatcher_->setSpriteRegistry(spriteRegistry_);
    }
}

void FrameContext::enterFrame(int frame) {
    beginSpritesForFrame(frame);
    initializeFrameScript(frame);
    dispatchBeginSprite();
    enteredChannels_.clear();
}

void FrameContext::beginSpritesForFrame(int frame) {
    auto spans = navigator().getActiveSprites(frame);
    std::ranges::sort(spans, [](const score::SpriteSpan& left, const score::SpriteSpan& right) {
        return left.channel() < right.channel();
    });

    for (const auto& span : spans) {
        const int channel = span.channel();
        if (activeChannels_.contains(channel)) {
            continue;
        }

        activeChannels_.insert(channel);
        enteredChannels_.insert(channel);
        ensureScoreSpriteState(frame, channel);

        for (const auto& behaviorRef : span.behaviors()) {
            if (spriteRegistry_ != nullptr) {
                spriteRegistry_->markScoreBehaviorChannel(channel);
            }
            (void)behaviorManager().createInstance(behaviorRef, channel);
        }
        logEvent("beginSprite");
    }
}

void FrameContext::ensureScoreSpriteState(int frame, int channel) {
    if (spriteRegistry_ == nullptr || file_ == nullptr) {
        return;
    }
    auto score = file_->scoreChunk();
    if (!score) {
        return;
    }
    const int frameIndex = frame - 1;
    for (const auto& entry : score->frameData().frameChannelData) {
        if (entry.frameIndex.value() == frameIndex && entry.channelIndex.value() == channel) {
            (void)spriteRegistry_->getOrCreate(channel, entry.data);
            return;
        }
    }
}

void FrameContext::endSpritesLeavingFrame(int oldFrame, int newFrame) {
    (void)oldFrame;
    const auto newActiveChannels = navigator().getActiveChannels(newFrame);
    std::vector<int> leaving;
    for (const int channel : activeChannels_) {
        if (newActiveChannels.contains(channel)) {
            continue;
        }
        if (spriteRegistry_ != nullptr) {
            auto state = spriteRegistry_->get(channel);
            if (state && state->isPuppet()) {
                logEvent("endSprite skipped for puppeted sprite");
                continue;
            }
        }
        leaving.push_back(channel);
    }

    for (const int channel : leaving) {
        eventDispatcher().dispatchSpriteEvent(channel, PlayerEvent::EndSprite);
        for (const auto& instance : behaviorManager().getInstancesForChannel(channel)) {
            if (instance) {
                instance->setEndSpriteCalled(true);
            }
        }
        activeChannels_.erase(channel);
        behaviorManager().removeInstancesForChannel(channel);
        if (spriteRegistry_ != nullptr) {
            spriteRegistry_->remove(channel);
        }
        logEvent("endSprite");
    }
}

void FrameContext::initializeFrameScript(int frame) {
    if (const auto* frameScript = navigator().getFrameScript(frame); frameScript != nullptr) {
        (void)behaviorManager().getOrCreateFrameScript(*frameScript, frame);
        logEvent("initializeFrameScript");
    } else {
        logEvent("initializeFrameScript: none");
    }
}

void FrameContext::dispatchToActorList(std::string_view handlerName) {
    if (actorListDispatcher_) {
        actorListDispatcher_(handlerName);
    }
}

void FrameContext::dispatchEvent(PlayerEvent event) {
    eventDispatcher().dispatchGlobalEvent(event);
    notifyEvent(event);
}

void FrameContext::dispatchBeginSprite() {
    for (const int channel : enteredChannels_) {
        for (const auto& instance : behaviorManager().getInstancesForChannel(channel)) {
            if (!instance || instance->isBeginSpriteCalled()) {
                continue;
            }
            eventDispatcher().dispatchBehaviorEvent(instance, PlayerEvent::BeginSprite);
            instance->setBeginSpriteCalled(true);
        }
    }

    auto frameInstance = behaviorManager().frameScriptInstance();
    if (frameInstance && !frameInstance->isBeginSpriteCalled()) {
        eventDispatcher().dispatchFrameAndMovieEvent(PlayerEvent::BeginSprite);
        frameInstance->setBeginSpriteCalled(true);
    }
}

bool FrameContext::hasScoreSpriteBehaviors(int frame) const {
    for (const auto& span : navigator().getActiveSprites(frame)) {
        if (!span.behaviors().empty()) {
            return true;
        }
    }
    return false;
}

bool FrameContext::hasFrameOneFrameScriptAndFrameTwoSpriteBehaviorStartupShape() const {
    return navigator().getFrameScript(1) != nullptr &&
           !hasScoreSpriteBehaviors(1) &&
           hasScoreSpriteBehaviors(2);
}

void FrameContext::logEvent(std::string_view message) const {
    if (debugEnabled_) {
        std::cout << "[FrameContext] " << message << '\n';
    }
}

void FrameContext::notifyEvent(PlayerEvent event) {
    if (eventListener_) {
        eventListener_(FrameEvent{event, currentFrame_});
    }
}

} // namespace libreshockwave::player::frame
